src/lib/gadgets/foreign-field.ts

```diff
@@ -0,0 +1,138 @@
+import { mod } from '../../bindings/crypto/finite_field.js';
+import { provableTuple } from '../../bindings/lib/provable-snarky.js';
+import { Field } from '../field.js';
+import { Gates, foreignFieldAdd } from '../gates.js';
+import { Tuple } from '../util/types.js';
+import { assert, exists, toVars } from './common.js';
+import { L, lMask, multiRangeCheck, twoL, twoLMask } from './range-check.js';
+
+export { ForeignField, Field3, Sign };
+
+/**
+ * A 3-tuple of Fields, representing a 3-limb bigint.
+ */
+type Field3 = [Field, Field, Field];
+type bigint3 = [bigint, bigint, bigint];
+type Sign = -1n | 1n;
+
+const ForeignField = {
+  add(x: Field3, y: Field3, f: bigint) {
+    return sum([x, y], [1n], f);
+  },
+  sub(x: Field3, y: Field3, f: bigint) {
+    return sum([x, y], [-1n], f);
+  },
+  sum,
+};
+
+/**
+ * computes x[0] + sign[0] * x[1] + ... + sign[n-1] * x[n] modulo f
+ *
+ * assumes that inputs are range checked, does range check on the result.
+ */
+function sum(x: Field3[], sign: Sign[], f: bigint) {
```

Comment:
`sum()` is exposed! it is actually the same as `sumchain` in OCaml, I renamed it because I think the "chain" aspect is more about the internal implementation and less important for users